python/lsst/meas/extensions/scarlet/io.py

```diff
@@ -472,3 +482,122 @@ def oldScarletToData(blend: Blend, psfCenter: tuple[int, int], xy0: Point2I):
+            'psf': model_psf,
+            'psfShape': psf_shape,
+            'blends': blends,
+        })
+
+    def _model_to_zip(self, model_data: scl.io.ScarletModelData) -> dict[str, Any]:
+        """Convert a ScarletModelData instance to a dictionary of files.
+
+        Parameters
+        ----------
+        model_data : `lsst.scarelt.lite.io.ScarletModelData`
+            ScarletModelData instance.
+
+        Returns
+        -------
+        zip_dict :
+            Dictionary mapping filenames to the desired file type.
+        """
+        json_model = model_data.as_dict()
+
+        data = {
+            str(blend_id): json.dumps(blend_data)
+            for blend_id, blend_data in json_model['blends'].items()
+        }
+        data.update({
+            'psf_shape': json.dumps(json_model['psfShape']),
+            'psf': json.dumps(json_model['psf']),
+        })
+        return data
+
+    def read_from_stream(
+        self, stream: BinaryIO | ResourceHandleProtocol, component: str | None = None, expected_size: int = -1
+    ) -> Any:
+        # Override of `FormatterV2.read_from_stream`.
+        if self.file_descriptor.parameters is not None and "blend_id" in self.file_descriptor.parameters:
+            filename = self.file_descriptor.parameters["blend_id"]
+            if isinstance(filename, Iterable):
+                filenames = [str(f) for f in filename]
+            else:
+                filenames = [str(filename)]
+            filenames += ['psf', 'psf_shape']
+        else:
+            filenames = None
+
+        with zipfile.ZipFile(stream, 'r') as zip_file:
+            if filenames is None:
+                filenames = [filename for filename in zip_file.namelist()]
+
+            unzipped_files = {}
+            for filename in filenames:
+                with zip_file.open(filename) as f:
+                    unzipped_files[filename] = from_json(f.read())
+
+            return self._build_model(unzipped_files)
+
+    def to_bytes(self, in_memory_dataset: Any) -> bytes:
+        in_memory_zip = BytesIO()
+        with zipfile.ZipFile(in_memory_zip, 'w') as zf:
+            zip_archive = self._model_to_zip(in_memory_dataset)
+            for filename, data in zip_archive.items():
+                zf.writestr(filename, data)
+        return in_memory_zip.getvalue()
+
+
+class ScarletModelDelegate(StorageClassDelegate):
+    """Delegate to extract a blend from an in-memory ScarletModelData object.
+    """
+    def can_accept(self, inMemoryDataset: Any) -> bool:
+        return isinstance(inMemoryDataset, scl.io.ScarletModelData)
+
+    def getComponent(self, composite: Any, componentName: str) -> Any:
+        raise AttributeError(f"Unsupported component: {componentName}")
+
+    def handleParameters(self, inMemoryDataset: Any, parameters: Mapping[str, Any] | None = None) -> Any:
+        if "blend_id" in parameters:
+            blend_id = parameters["blend_id"]
+            if isinstance(blend_id, Iterable):
```

Comment:
you can use `blend_ids = lsst.utils.iteration.ensure_iterable(blend_id)` to clean up this code a bit (and the code above).